plugins/animate/squeezimize.hpp

```diff
@@ -0,0 +1,388 @@
+#include <wayfire/signal-definitions.hpp>
+#include <wayfire/toplevel-view.hpp>
+#include <wayfire/window-manager.hpp>
+#include <wayfire/view-transform.hpp>
+#include <wayfire/txn/transaction-manager.hpp>
+#include <wayfire/render-manager.hpp>
+#include <wayfire/scene-render.hpp>
+#include <wayfire/util/duration.hpp>
+#include <glm/gtc/matrix_transform.hpp>
+
+#include "animate.hpp"
+
+
+static std::string sqeezimize_transformer_name = "animation-squeezimize";
+
+wf::option_wrapper_t<wf::animation_description_t> squeezimize_duration{"animate/squeezimize_duration"};
+
+static const char *squeeze_vert_source =
+    R"(
+#version 100
+
+attribute mediump vec2 position;
+attribute mediump vec2 uv_in;
+
+uniform mat4 matrix;
+
+varying highp vec2 uv;
+
+void main() {
+    uv = uv_in;
+    gl_Position = matrix * vec4(position, 0.0, 1.0);
+}
+)";
+
+static const char *squeeze_frag_source =
+    R"(
+#version 100
+@builtin_ext@
+@builtin@
+
+precision mediump float;
+
+varying highp vec2 uv;
+uniform mediump float progress;
+uniform mediump vec4 src_box;
+uniform mediump vec4 target_box;
+uniform int upward;
+
+void main()
+{
+    float y;
+    vec2 uv_squeeze;
+    float inv_w = 1.0 / (src_box.z - src_box.x);
+    float inv_h = 1.0 / (src_box.w - src_box.y);
+    float progress_pt_one = clamp(progress, 0.0, 0.5) * 2.0;
+    float progress_pt_two = (clamp(progress, 0.5, 1.0) - 0.5) * 2.0;
+
+    uv_squeeze.x = (uv.x * inv_w) - (inv_w - 1.0);
+    uv_squeeze.x += inv_w - inv_w * src_box.z;
+    uv_squeeze.y = (uv.y * inv_h) - (inv_h - 1.0);
+    uv_squeeze.y += inv_h * src_box.y;
+
+    if (upward == 1)
+    {
+        y = uv.y;
+        uv_squeeze.y += -progress_pt_two * (inv_h - target_box.w);
+    } else
+    {
+        y = 1.0 - uv.y;
+        uv_squeeze.y -= -progress_pt_two * (src_box.y + target_box.y + target_box.w);
+    }
+
+    float sigmoid = 1.0 / (1.0 + pow(2.718, -((y * inv_h) * 6.0 - 3.0)));
+    sigmoid *= progress_pt_one * (src_box.x - target_box.x);
+
+    uv_squeeze.x += sigmoid * inv_w;
+    uv_squeeze.x *= (y * (1.0 / (target_box.z - target_box.x)) * progress_pt_one) + 1.0;
+
+    if (uv_squeeze.x < 0.0 || uv_squeeze.y < 0.0 ||
+        uv_squeeze.x > 1.0 || uv_squeeze.y > 1.0)
+    {
+        discard;
+    }
+
+    gl_FragColor = get_pixel(uv_squeeze);
+}
+)";
+
+namespace wf
+{
+namespace squeezimize
+{
+using namespace wf::scene;
+using namespace wf::animation;
+class squeezimize_animation_t : public duration_t
+{
+  public:
+    using duration_t::duration_t;
+    timed_transition_t squeeze{*this};
+};
+class squeezimize_transformer : public wf::scene::view_2d_transformer_t
+{
+  public:
+    wayfire_view view;
+    OpenGL::program_t program;
+    wf::output_t *output;
+    wf::geometry_t minimize_target;
+    wf::geometry_t animation_geometry;
+    squeezimize_animation_t progression{squeezimize_duration};
+
+    class simple_node_render_instance_t : public wf::scene::transformer_render_instance_t<transformer_base_node_t>
```

Comment:
You can specify the actual transformer type here `wf::scene::transformer_render_instance_t<squeezimize_transformer>` as a base class. Then you can remove `squeezimize_transformer *self` that you have a few lines below, because the base class will already store this pointer for you: https://github.com/WayfireWM/wayfire/blob/0daf5c9a1f52383192d53bb20ad0d770581413f9/src/api/wayfire/view-transform.hpp#L148

Reply:
Interesting, thanks for the tip.